src/Compilers/Core/Portable/Collections/Rope.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A representation of a string of characters that requires O(1) extra space to concatenate two ropes.
+    /// </summary>
+    internal abstract class Rope
+    {
+        public static readonly Rope Empty = ForString("");
+        public abstract override string ToString();
+        public abstract int Length { get; }
+        protected abstract IEnumerable<char> GetChars();
```

Comment:
Did you consider using a `Span<char>` here given this type is meant to help with perf scenarios?